pages/CitationStyleLanguageHandler.inc.php

```diff
@@ -105,54 +114,43 @@ public function _setupRequest($args, $request)
-            $issue = $issue->getJournalId() == $context->getId() ? $issue : null;
-            $this->issue = $issue;
-        }
-
         // Disallow access to unpublished submissions, unless the user is a
         // journal manager or an assigned subeditor or assistant. This ensures the
-        // article preview will work for those who can see it.
-        if (!$this->issue || !$this->issue->getPublished() || $this->article->getStatus() != PKPSubmission::STATUS_PUBLISHED) {
-            $userCanAccess = false;
-
-            if ($user && $user->hasRole([Role::ROLE_ID_SUB_EDITOR, Role::ROLE_ID_ASSISTANT], $context->getId())) {
-                $isAssigned = false;
-                $userGroupDao = DAORegistry::getDAO('UserGroupDAO');
-                $stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO');
-                $assignments = $stageAssignmentDao->getBySubmissionAndStageId($this->article->getId());
-                foreach ($assignments as $assignment) {
-                    if ($assignment->getUser()->getId() !== $user->getId()) {
-                        continue;
-                    }
-                    $userGroup = $userGroupDao->getById($assignment->getUserGroupId($context->getId()));
-                    if (in_array($userGroup->getRoleId(), [Role::ROLE_ID_SUB_EDITOR, Role::ROLE_ID_ASSISTANT])) {
-                        $userCanAccess = true;
-                        break;
-                    }
-                }
+        // submission preview will work for those who can see it.
+        if ($this->isSubmissionUnpublished($this->issue, $this->submission)) {
+            if (!$this->canUserAccess($context, $user)) {
+                $request->getDispatcher()->handle404();
             }
+        }
+    }
 
-            if ($user && $user->hasRole(Role::ROLE_ID_MANAGER, $context->getId())) {
-                $userCanAccess = true;
+    protected function canUserAccess($context, $user) {
+        if ($user && $user->hasRole([Role::ROLE_ID_SUB_EDITOR, Role::ROLE_ID_ASSISTANT], $context->getId())) {
+            $userGroupDao = DAORegistry::getDAO('UserGroupDAO');
+            $stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO');
+            $assignments = $stageAssignmentDao->getBySubmissionAndStageId($this->submission->getId());
+            foreach ($assignments as $assignment) {
+                if ($assignment->getUser()->getId() == $user->getId()) {
+                    continue;
+                }
+                $userGroup = $userGroupDao->getById($assignment->getUserGroupId($context->getId()));
+                if (in_array($userGroup->getRoleId(), [Role::ROLE_ID_SUB_EDITOR, Role::ROLE_ID_ASSISTANT])) {
+                    return true;
+                }
             }
+        }
 
-            if (!$userCanAccess) {
-                $request->getDispatcher()->handle404();
-            }
+        if ($user && $user->hasRole(Role::ROLE_ID_MANAGER, $context->getId())) {
```

Comment:
We're trying to reduce the use of `$user->hasRole()` because it requires an extra database lookup. Because it is the request user we are looking up, we may be able to check the authorized roles.

From within `_setupRequest()`:

```php
$userRoles = $this->getAuthorizedContextObject(ASSOC_TYPE_USER_ROLES);
$this->canUserAccess($context, $user, $userRoles);
```

And in `canUserAccess()`:

```php
if ($user && in_array(Role::ROLE_ID_MANAGER, $userRoles)) {
    ...
}
```

To replace `hasRole()` on this line and above at line 135.